### ISSUE TITLE
Some missed metrics with Alloy 1.4.x on Windows Server 2022 Datacenter

### ISSUE THREAD

**Issue by halt-avmc** [grafana/alloy#1845]
### What's wrong?

Alloy does not export windows metrics, at least for

- net
- logical_disk
- phisical_disk

while no errors or timeout are reported:

```
# HELP windows_exporter_build_info A metric with a constant '1' value labeled by version, revision, branch, goversion from which windows_exporter was built, and the goos and goarch for the build.
# TYPE windows_exporter_build_info gauge
windows_exporter_build_info{branch="HEAD",goarch="amd64",goos="windows",goversion="go1.22.7",revision="f59e491e5",tags="builtinassets",version="v1.4.2"} 1
# HELP windows_exporter_collector_duration_seconds windows_exporter: Duration of a collection.
# TYPE windows_exporter_collector_duration_seconds gauge
windows_exporter_collector_duration_seconds{collector="logical_disk"} 0
windows_exporter_collector_duration_seconds{collector="net"} 0
windows_exporter_collector_duration_seconds{collector="physical_disk"} 0
# HELP windows_exporter_collector_success windows_exporter: Whether the collector was successful.
# TYPE windows_exporter_collector_success gauge
windows_exporter_collector_success{collector="logical_disk"} 1
windows_exporter_collector_success{collector="net"} 1
windows_exporter_collector_success{collector="physical_disk"} 1
# HELP windows_exporter_collector_timeout windows_exporter: Whether the collector timed out.
# TYPE windows_exporter_collector_timeout gauge
windows_exporter_collector_timeout{collector="logical_disk"} 0
windows_exporter_collector_timeout{collector="net"} 0
windows_exporter_collector_timeout{collector="physical_disk"} 0
# HELP windows_exporter_perflib_snapshot_duration_seconds Duration of perflib snapshot capture
# TYPE windows_exporter_perflib_snapshot_duration_seconds gauge
windows_exporter_perflib_snapshot_duration_seconds 0.0014556
```
and that is all what is shown, there are no metrics at all of
```
windows_net_xxx
windows_logical_disk_xxx
windows_phisical_disk_xxx
```
while other windows metrics (like cpu, system, etc.) are actually exported normally

FYI: **Alloy v1.3.4** and **windows_exporter v0.29.1** are working correctly

### Steps to reproduce

On Windows Server 2022 Datacenter install Grafana Alloy 1.3.4, use attached config, open in browser [metrics path](http://localhost:12345/api/v0/component/prometheus.exporter.windows.testing/metrics) and see all of them are present:

```
# HELP windows_exporter_build_info A metric with a constant '1' value labeled by version, revision, branch, goversion from which windows_exporter was built, and the goos and goarch for the build.
# TYPE windows_exporter_build_info gauge
windows_exporter_build_info{branch="HEAD",goarch="amd64",goos="windows",goversion="go1.22.5",revision="88e779887",tags="builtinassets",version="v1.3.4"} 1
# HELP windows_exporter_collector_duration_seconds windows_exporter: Duration of a collection.
# TYPE windows_exporter_collector_duration_seconds gauge
windows_exporter_collector_duration_seconds{collector="logical_disk"} 0.0320621
windows_exporter_collector_duration_seconds{collector="net"} 4.16e-05
windows_exporter_collector_duration_seconds{collector="physical_disk"} 0.0003873
# HELP windows_exporter_collector_success windows_exporter: Whether the collector was successful.
# TYPE windows_exporter_collector_success gauge
windows_exporter_collector_success{collector="logical_disk"} 1
windows_exporter_collector_success{collector="net"} 1
windows_exporter_collector_success{collector="physical_disk"} 1
# HELP windows_exporter_collector_timeout windows_exporter: Whether the collector timed out.
# TYPE windows_exporter_collector_timeout gauge
windows_exporter_collector_timeout{collector="logical_disk"} 0
windows_exporter_collector_timeout{collector="net"} 0
windows_exporter_collector_timeout{collector="physical_disk"} 0
# HELP windows_exporter_perflib_snapshot_duration_seconds Duration of perflib snapshot capture
# TYPE windows_exporter_perflib_snapshot_duration_seconds gauge
windows_exporter_perflib_snapshot_duration_seconds 0.006464
# HELP windows_logical_disk_avg_read_requests_queued Average number of read requests that were queued for the selected disk during the sample interval (LogicalDisk.AvgDiskReadQueueLength)
# TYPE windows_logical_disk_avg_read_requests_queued gauge
windows_logical_disk_avg_read_requests_queued{volume="C:",volume_name="Local Disk"} 2585.9816283

and a whole list of related metrics below ...
```

Upgrade or do a clean install of Alloy 1.4.0 / 1.4.1 / 1.4.2 with the same config and see mentioned metrics are absent

### System information

Windows Server 2022 Datacenter, version 21H2, OS build 20348.887

### Software version

Grafana Alloy 1.4.0, Alloy 1.4.1, Alloy 1.4.2

### Configuration

config.alloy:
```
logging {
	level = "info"
}

prometheus.exporter.windows "node" {
  enabled_collectors = ["cpu", "cs", "os", "system", "mssql"]
}

prometheus.exporter.windows "testing" {
  enabled_collectors = ["net", "physical_disk", "logical_disk"]
}
```

========================================================================================

windows_exporter config.yaml for reference

```
log:
  level: warn
collectors:
  enabled: logical_disk,physical_disk,net
collector:
  service:
    include: windows_exporter
```

### Logs

From Windows EventLog

```
level=info msg="starting program" command="C:/Program Files/GrafanaLabs/Alloy/alloy-windows-amd64.exe run C:\\Program Files\\GrafanaLabs\\Alloy\\config.alloy --storage.path=C:\\ProgramData\\GrafanaLabs\\Alloy\\data --disable-reporting --server.http.enable-pprof=false --server.http.listen-addr=0.0.0.0:12345"

ts=2024-10-08T08:15:47.6325903Z level=info msg="no peer discovery configured: both join and discover peers are empty" service=cluster
ts=2024-10-08T08:15:47.6325903Z level=info msg="starting complete graph evaluation" controller_path=/ controller_id="" trace_id=d9e5080bcd73b22ef5342ccc0903856b
ts=2024-10-08T08:15:47.6325903Z level=info msg="finished node evaluation" controller_path=/ controller_id="" trace_id=d9e5080bcd73b22ef5342ccc0903856b node_id=livedebugging duration=0s
ts=2024-10-08T08:15:47.6325903Z level=info msg="enabled windows_exporter collectors" component_path=/ component_id=prometheus.exporter.windows.testing collectors=logical_disk,net,physical_disk
ts=2024-10-08T08:15:47.6325903Z level=info msg="finished node evaluation" controller_path=/ controller_id="" trace_id=d9e5080bcd73b22ef5342ccc0903856b node_id=prometheus.exporter.windows.testing duration=62.4166ms
ts=2024-10-08T08:15:47.6325903Z level=info msg="enabled windows_exporter collectors" component_path=/ component_id=prometheus.exporter.windows.node collectors=cpu,cs,mssql,os,system
ts=2024-10-08T08:15:47.6325903Z level=info msg="finished node evaluation" controller_path=/ controller_id="" trace_id=d9e5080bcd73b22ef5342ccc0903856b node_id=prometheus.exporter.windows.node duration=0s
ts=2024-10-08T08:15:47.6325903Z level=info msg="finished node evaluation" controller_path=/ controller_id="" trace_id=d9e5080bcd73b22ef5342ccc0903856b node_id=logging duration=0s
ts=2024-10-08T08:15:47.6346102Z level=info msg="finished node evaluation" controller_path=/ controller_id="" trace_id=d9e5080bcd73b22ef5342ccc0903856b node_id=remotecfg duration=2.0199ms
ts=2024-10-08T08:15:47.6346102Z level=info msg="applying non-TLS config to HTTP server" service=http
ts=2024-10-08T08:15:47.6346102Z level=info msg="finished node evaluation" controller_path=/ controller_id="" trace_id=d9e5080bcd73b22ef5342ccc0903856b node_id=http duration=0s
ts=2024-10-08T08:15:47.6346102Z level=info msg="finished node evaluation" controller_path=/ controller_id="" trace_id=d9e5080bcd73b22ef5342ccc0903856b node_id=cluster duration=0s
ts=2024-10-08T08:15:47.6346102Z level=info msg="finished node evaluation" controller_path=/ controller_id="" trace_id=d9e5080bcd73b22ef5342ccc0903856b node_id=ui duration=0s
ts=2024-10-08T08:15:47.6346102Z level=info msg="finished node evaluation" controller_path=/ controller_id="" trace_id=d9e5080bcd73b22ef5342ccc0903856b node_id=otel duration=0s
ts=2024-10-08T08:15:47.6346102Z level=info msg="finished node evaluation" controller_path=/ controller_id="" trace_id=d9e5080bcd73b22ef5342ccc0903856b node_id=labelstore duration=0s
ts=2024-10-08T08:15:47.6346102Z level=info msg="finished node evaluation" controller_path=/ controller_id="" trace_id=d9e5080bcd73b22ef5342ccc0903856b node_id=tracing duration=0s

ts=2024-10-08T08:15:47.6325903Z level=info "boringcrypto enabled"=false

ts=2024-10-08T08:15:47.6346102Z level=info msg="finished complete graph evaluation" controller_path=/ controller_id="" trace_id=d9e5080bcd73b22ef5342ccc0903856b duration=64.4365ms
ts=2024-10-08T08:15:47.6346102Z level=info msg="scheduling loaded components and services"

ts=2024-10-08T08:15:47.6351612Z level=info msg="starting cluster node" service=cluster peers_count=0 peers="" advertise_addr=127.0.0.1:12345
ts=2024-10-08T08:15:47.6353867Z level=info msg="peers changed" service=cluster peers_count=1 peers=SQL1

ts=2024-10-08T08:15:47.636747Z level=info msg="now listening for http traffic" service=http addr=0.0.0.0:12345
```

**Comment by ptodev** [comment]
Hi, @halt-avmc 👋  Thank you for letting us know about this issue. I was able to reproduce it by following your instructions.

Hi @jkroepke 👋  This looks like a bug in the windows exporter. When the `exclude` [argument](https://grafana.com/docs/alloy/latest/reference/components/prometheus/prometheus.exporter.windows/#network-block) is set to `""`, the `c.config.NicExclude.MatchString(nic.Name)` [expression](https://github.com/prometheus-community/windows_exporter/blob/v0.27.3/pkg/collector/net/net.go#L251C6-L251C47) always returns true. I suppose the Windows exporter should be changed to not check `c.config.NicExclude` and `c.config.NicInclude` when the regex is an empty string?

When I change the config to this, I see the missing network metrics:
```
prometheus.exporter.windows "testing" {
  enabled_collectors = ["net", "physical_disk", "logical_disk"]
  network {
    exclude = "test"
  }
}
```

**Comment by jkroepke** [comment]
@ptodev That's thats something what I recently notice while integrating tests at windows_exporter. I will do a backport soon.

**Comment by Nachtfalkeaw** [comment]
Hello,

unfortunately I missed that issue and I created a new one:
https://github.com/grafana/alloy/issues/1860

It is the same issue with "process" - If I add something to the "exclude" then it is working.

From my tests this started with Grafana Alloy 1.4.x and did not appear in 1.3.3

**Comment by ptodev** [comment]
Thanks, @jkroepke! Would you prefer to rollback the version of Windows Exporter in Alloy? I'm not sure how long the fix would take, and this seems like it could be breaking a lot of users.

**Comment by halt-avmc** [comment]
@ptodev thanks for checking

I've checked your config with **Alloy 1.4.2** and can confirm it can be used as workaround. But it looks like all modules of interest must use "exclude" clause. Because using your example config (with "network" block)  - only `windows_net_xxx` metrics are began to expose. Other was still unavailable

So, to actually expose all mentioned metrics I made this config:

```
logging {
	level = "info"
}

prometheus.exporter.windows "testing" {
  enabled_collectors = ["net", "physical_disk", "logical_disk"]
  network {
    exclude = "test" // Workaround for Alloy 1.4.2
  }
  logical_disk {
    exclude = "test" // Workaround for Alloy 1.4.2
  }
  physical_disk {
    exclude = "test" // Workaround for Alloy 1.4.2
  }
}
```
With this config all metrics are exposed correctly.

It looks like it can affect [another metrics too](https://github.com/grafana/alloy/issues/1845#issuecomment-2404194470). I assume it affects exporting of all metrics from `prometheus.exporter.windows` which can have "exclude" clause in config.

BTW: in documentation about [prometheus.windows.exporter](https://grafana.com/docs/alloy/latest/reference/components/prometheus/prometheus.exporter.windows/#blocks) there is no mention about availability of "physical_disk" block, but it actually exists and used by Alloy
